src/app/core/config/config.service.ts

```diff
@@ -48,4 +49,88 @@ export class ConfigService extends LatestEntityLoader<Config> {
     }
     return matchingConfigs;
   }
+
+  private applyMigrations(config: Config): Config {
+    const migrations: ConfigMigration[] = [
+      migrateFormHeadersIntoFieldGroups,
+      migrateFormFieldConfigView2ViewComponent,
+    ];
+
+    const newConfig = JSON.parse(JSON.stringify(config), (_that, rawValue) => {
+      let configPart = rawValue;
+      for (const migration of migrations) {
+        if (migration.filter(_that, configPart)) {
+          configPart = migration.transform(_that, configPart);
+        }
+      }
+      return configPart;
+    });
+
+    return newConfig;
+  }
 }
+
+/**
+ * A ConfigMigration is checked during a full JSON.parse using a reviver function.
+ * If the filter returns true, the transform is executed on that config part.
+ * Multiple migrations are chained and can transform the same config part one after the other.
+ */
+interface ConfigMigration {
+  /** filter returning true for a given config part if transform should be executed on it */
+  filter: (key: string, configPart: any) => boolean;
+  transform: (key: string, configPart: any) => any;
+}
+
+/**
+ * Transform legacy "view:...Form" config format to have form field group headers with the fields rather than as separate array.
+ */
+const migrateFormHeadersIntoFieldGroups: ConfigMigration = {
+  filter: (key, configPart) =>
+    configPart?.component === "Form" && configPart?.config?.cols,
+
+  transform: (key, configPart) => {
+    const formConfig = configPart.config;
+
+    // change .cols and .headers into .fieldGroups
+    const newFormConfig = { ...formConfig };
+    delete newFormConfig.cols;
+    delete newFormConfig.headers;
+
+    newFormConfig.fieldGroups = formConfig.cols?.map(
+      (colGroup) => ({ fields: colGroup }) as FieldGroup,
+    );
+    if (formConfig.headers) {
+      newFormConfig.fieldGroups.forEach((group, i) => {
+        if (formConfig.headers[i]) {
+          group.header = formConfig.headers[i];
+        }
+      });
+    }
+
+    configPart.config = newFormConfig;
+    return configPart;
+  },
+};
+
+const migrateFormFieldConfigView2ViewComponent: ConfigMigration = {
```

Comment:
I think for this kind of migration it would be much simpler to just use search and replace. I don't see "view" and "edit" used anywhere else in our config.

Reply:
while that might work well, I would like to avoid a risky approach - there may well be some user defined pieces that accidentally match. Let's do this properly.